client/src/main/scala/com.gu.contentapi.client/GuardianContentClient.scala

```diff
@@ -90,77 +95,32 @@ trait ContentApiClientLogic {
-    }
-
-  def getResponse(removedContentQuery: RemovedContentQuery)(implicit context: ExecutionContext): Future[RemovedContentResponse] =
-    fetchResponse(removedContentQuery) map { response =>
-      ThriftDeserializer.deserialize(response, RemovedContentResponse)
-    }
-
-  def getResponse(videoStatsQuery: VideoStatsQuery)(implicit context: ExecutionContext): Future[VideoStatsResponse] =
-    fetchResponse(videoStatsQuery) map { response =>
-      ThriftDeserializer.deserialize(response, VideoStatsResponse)
-    }
-
-  def getResponse(atomsQuery: AtomsQuery)(implicit context: ExecutionContext): Future[AtomsResponse] =
-    fetchResponse(atomsQuery) map { response =>
-      ThriftDeserializer.deserialize(response, AtomsResponse)
-    }
-
-  def getResponse(recipesQuery: RecipesQuery)(implicit context: ExecutionContext): Future[AtomsResponse] =
-    fetchResponse(recipesQuery) map { response =>
-      ThriftDeserializer.deserialize(response, AtomsResponse)
-    }
-
-  def getResponse(reviewsQuery: ReviewsQuery)(implicit context: ExecutionContext): Future[AtomsResponse] =
-    fetchResponse(reviewsQuery) map { response =>
-      ThriftDeserializer.deserialize(response, AtomsResponse)
-    }
-
-  def getResponse(gameReviewsQuery: GameReviewsQuery)(implicit context: ExecutionContext): Future[AtomsResponse] =
-    fetchResponse(gameReviewsQuery) map { response =>
-      ThriftDeserializer.deserialize(response, AtomsResponse)
-    }
-
-  def getResponse(restaurantReviewsQuery: RestaurantReviewsQuery)(implicit context: ExecutionContext): Future[AtomsResponse] =
-    fetchResponse(restaurantReviewsQuery) map { response =>
-      ThriftDeserializer.deserialize(response, AtomsResponse)
+  private def paginate2[Q <: PaginatedApiQuery[Q], RR](q: Q, f: RR => Future[Unit])(r: RR)(
+    implicit
+    decoder: Decoder[Q] { type R = RR },
+    meta: MetaResult[RR],
+    context: ExecutionContext): Future[Unit] =
+    f(r).flatMap { _ =>
+      (meta.isLastPage(r), meta.getResults(r).lastOption.map(meta.getId)) match {
+        case (false, Some(id)) => getResponse(next(q, id)).flatMap(paginate2(q, f)(_))
+        case _                 => Future.successful(())
+      }
     }
 
-  def getResponse(filmReviewsQuery: FilmReviewsQuery)(implicit context: ExecutionContext): Future[AtomsResponse] =
-    fetchResponse(filmReviewsQuery) map { response =>
-      ThriftDeserializer.deserialize(response, AtomsResponse)
-    }
+  /* Exposed API */
 
-  def getResponse(storiesQuery: StoriesQuery)(implicit context: ExecutionContext): Future[StoriesResponse] =
-    fetchResponse(storiesQuery) map { response =>
-      ThriftDeserializer.deserialize(response, StoriesResponse)
-    }
+  def getResponse[Q <: ContentApiQuery](query: Q)(
+    implicit 
+    decoder: Decoder[Q],
+    context: ExecutionContext): Future[decoder.R] =
+    fetchResponse(query) map decoder.decode
+
+  def paginate[Q <: PaginatedApiQuery[Q], RR: MetaResult](q: Q)(f: RR => Future[Unit])(
```

Comment:
It took me some time to notice the different operators in this expression, and particularly the `implicit MetaResult[RR]`. Since this function already declares an implicit parameter list, I feel that it would be more intuitive to have the `implicit MetaResult[RR]` declared in that list.